api/tools/src/main/java/org/pmiops/workbench/tools/ManageClusters.java

```diff
@@ -0,0 +1,170 @@
+        .sorted(Comparator.comparing(c -> Instant.parse(c.getCreatedDate())))
+        .forEachOrdered((c) -> {
+              System.out.println(formatTabular(c));
+              count.getAndIncrement();
+            });
+    System.out.println(String.format("listed %d clusters", count.get()));
+  }
+
+  private static void deleteClusters(Instant oldest, Set<String> ids, boolean dryRun)
+      throws IOException, ApiException {
+    Set<String> remaining = new HashSet<>(ids);
+    String dryMsg = dryRun? "[DRY RUN]: would have... " : "";
+
+    AtomicInteger deleted = new AtomicInteger();
+    ClusterApi api = newApiClient();
+    api.listClusters(null, false).stream()
+        .sorted(Comparator.comparing(c -> Instant.parse(c.getCreatedDate())))
+        .filter((c) -> {
+              Instant createdDate = Instant.parse(c.getCreatedDate());
+              if (oldest != null && createdDate.isAfter(oldest)) {
+                return false;
+              }
+              if (!ids.isEmpty() && !ids.contains(clusterId(c))) {
+                return false;
+              }
+              return true;
+            })
+        .forEachOrdered((c) -> {
+              String cid = clusterId(c);
+              if (!dryRun) {
+                try {
+                  api.deleteCluster(c.getGoogleProject(), c.getClusterName());
+                } catch (ApiException e) {
+                  log.log(Level.SEVERE, "failed to deleted cluster " + cid, e);
+                  return;
+                }
+              }
+              remaining.remove(cid);
+              deleted.getAndIncrement();
+              System.out.println(dryMsg + "deleted cluster: " + formatTabular(c));
+            });
+    if (!remaining.isEmpty()) {
+      log.log(Level.SEVERE, "failed to find/delete clusters: {1}",
+          new Object[]{Joiner.on(", ").join(remaining)});
+    }
+    System.out.println(String.format("%sdeleted %d clusters", dryMsg, deleted.get()));
+  }
+
+  @Bean
+  public CommandLineRunner run() {
+    return (args) -> {
+      if (args.length < 1) {
+        throw new IllegalArgumentException("must specify a command 'list' or 'delete'");
+      }
+      String cmd = args[0];
+      args = Arrays.copyOfRange(args, 1, args.length);
+      switch (cmd) {
+        case "list":
+          if (args.length > 0) {
+            throw new IllegalArgumentException("Expected 0 args. Got " + Arrays.asList(args));
+          }
+          listClusters();
+          return;
+
+        case "delete":
+          // User-friendly command-line parsing is done in devstart.rb, so we do only simple
+          // positional argument parsing here.
+          if (args.length != 3) {
+            throw new IllegalArgumentException(
+                "Expected 3 args (min_age, ids, dry_run). Got " + Arrays.asList(args));
+          }
+          Duration age = Duration.ZERO;
+          if (!args[0].isEmpty()) {
+            age = Duration.ofDays(Long.parseLong(args[0]));
```

Comment:
If age is zero... shouldn't oldest be set to null?

Reply:
Felt slightly cleaner to me to avoid the sentinel value null checks and just set the oldest to effectively "now()", i.e. no age restriction.